### PR TITLE
fix: parsing the header of confluent protobuf message incorrectly

### LINKED ISSUE #14773: bug(source): schema registry header of protobuf is not skipped properly
> This is skipping `i` **bytes** rather than skipping `i` **variable-length zig-zag encoded integers**. ~~For `i` within `1..=63`, the encoding outputs a single byte so this happens to work.~~
> 
> > prefixed by the length of the array (which is also variable length, Zigzag encoded)
> 
> `i` itself is also encoded so the decoded length shall be `i / 2`.

_Originally posted by @xiangjinwu in https://github.com/risingwavelabs/risingwave/pull/14057#discussion_r1464511298_

> Got it, so the correct implementation is
> ```
>         Some(i) => Ok(&remained[(1 + *i / 2) as usize..]),
> ```

_Originally posted by @Rossil2012 in https://github.com/risingwavelabs/risingwave/pull/14057#discussion_r1464561640_

### Impact

In the common case where a message is defined at top level (i.e. not nested within another message's definition), and the message index is <= 63, the current implement happens to work correctly because `i == 2` implies `i == 1 + i/2`.

Example: `UserWithNewType` in the PR is at top level with index 2, so its indexes array `[2]` would be encoded as `02 04` (`02` for the length `1` and `04` for the element `2`). Skipping `02` (first byte read as `i`) bytes happens to skip the encoded sequence exactly.

### PR DESCRIPTION
**Pull request by Rossil2012**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Resolve #14773.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by tabVersion — APPROVED** [review]
LGTM

share the same impl with confluent https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java#L159

File: src/connector/src/parser/protobuf/parser.rs [hunk 1]
```diff
@@ -524,21 +524,55 @@ fn protobuf_type_mapping(
     Ok(t)
 }
 
+/// A port from the implementation of confluent's Variant Zig-zag deserialization.
```
Comment (xiangjinwu) [inline src/connector/src/parser/protobuf/parser.rs:527]: It is not the word vari**a**nt, but actually var(iable length) int(eger).

File: src/connector/src/parser/protobuf/parser.rs [hunk 2]
```diff
@@ -1106,4 +1140,48 @@ mod test {
...
+        assert_eq!(len, 5);
+
+        // 6. Maximum negative number
+        let buffer = vec![0xFF, 0xFF, 0xFF, 0xFF, 0x0F];
+        let (value, len) = decode_variant_zigzag(&buffer).unwrap();
+        assert_eq!(value, i32::MIN);
+        assert_eq!(len, 5);
+
+        // 7. Invalid input (more than 5 bytes)
+        let buffer = vec![0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF];
```
Comment (xiangjinwu) [inline src/connector/src/parser/protobuf/parser.rs:1183]: What if input is `[0xFF, 0xFF, 0xFF, 0xFF, 0x7F]`?

Comment (Rossil2012) [reply src/connector/src/parser/protobuf/parser.rs:1183]: The result is i32::MIN, and the overflowed bits will be ignored. The behavior is the same as confluent's implementation.

Comment (Rossil2012) [reply src/connector/src/parser/protobuf/parser.rs:1183]: You can test the original impl here.
https://www.sololearn.com/en/compiler-playground/ccEr3Q77KgqB

**Review by xiangjinwu — APPROVED** [review]
Thanks!

**Pull request merged**